File: Problems/0475_Heaters/Project_CS/Heaters.cs
using System;
using System.Linq;

public class Solution
{
    public int FindRadius(int[] houses, int[] heaters)
    {
            houses = houses.OrderBy(x => x).ToArray();
            heaters = heaters.OrderBy(x => x).ToArray();

            int ret = 0;
            for (int i = 0, j = 0; i < heaters.Length && j < houses.Length;)
            {
                if (houses[j] <= heaters[i] && i == 0)
                {
                    ret = Math.Max(ret, heaters[i] - houses[j]);
                    j++;
                }
                else if (houses[j] >= heaters[i] && i == heaters.Length - 1)
                {
                    ret = Math.Max(ret, houses[j] - heaters[i]);
                    j++;
                }
                else if (i!= heaters.Length - 1 && houses[j] >= heaters[i] && houses[j] <= heaters[i + 1])
                {
                    ret = Math.Max(ret, Math.Min(houses[j] - heaters[i], heaters[i + 1] - houses[j]));
                    j++;
                }
                else
                {
                    i++;
                }
            }

            return ret;
        
    }
    public int[] str_to_int_array(string flds)
    {
        if (flds.Length <= 0)
            return null;

        string[] temp = flds.Split(',');
        int[] nums = new int[temp.Length];

        for (int i = 0; i < temp.Length; ++i)
        {
            nums[i] = int.Parse(temp[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length / 2; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string temp = args.Replace("[[","").Replace("]]","").Trim();
        string[] flds = temp.Split(new string[] {"],["}, StringSplitOptions.None);
        int[] houses = str_to_int_array(flds[0]);
        int[] heaters = str_to_int_array(flds[1]);

        Console.WriteLine("houses = " + output_int_array(houses));
        Console.WriteLine("heaters = " + output_int_array(heaters));
        
        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = FindRadius(houses, heaters);
        Console.WriteLine("result = " + result.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

File: Problems/0475_Heaters/Project_Python3/Heaters.py
# coding: utf-8

import os
import sys
import time

class Solution:
    def findRadius(self, houses, heaters):
        """
        :type houses: List[int]
        :type heaters: List[int]
        :rtype: int
        """
        houses.sort()
        heaters.sort()
        radius = 0
        i = 0
        for house in houses:
            while i < len(heaters) and heaters[i] < house:
                i += 1
            if i == 0:
                radius = max(radius, heaters[i] - house)
            elif i == len(heaters):
                return max(radius, houses[-1] - heaters[-1])
            else:
                radius = max(radius, min(heaters[i]-house, house-heaters[i-1]))
        return radius

    def findRadius_work(self, houses, heaters):
        """
        :type houses: List[int]
        :type heaters: List[int]
        :rtype: int
        """
        radius_min = [sys.maxsize]*len(houses)
        for i in range(len(houses)):
            for he in heaters:
                radius = abs(he - houses[i])
                if radius < radius_min[i]:
                    radius_min[i] = radius
        radius = 0
        for temp in radius_min:
            if temp > radius:
                radius = temp
        return radius

def str_to_int_array(flds):
    if len(flds) <= 0:
        return None
    temp = flds.split(",")
    nums = [0]*len(temp)
    for i in range(len(temp)):
        nums[i] = int(temp[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    var_str = temp.replace("[[","").replace("]]","").rstrip()
    flds = var_str.split("],[")
    houses = str_to_int_array(flds[0])
    heaters = str_to_int_array(flds[1])

    time0 = time.time()

    sl = Solution()
    result = sl.findRadius(houses, heaters)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()
